File: Src/Icm.Core.Tests/Reflection/MethodInfoExtensionsTest.vb
Imports Icm.Reflection

<TestFixture()>
Public Class MethodInfoExtensionsTest

    <Test()>
    Public Sub GetAttributesTest()
        Dim mi = GetType(ClassB).GetMethod("Routine")

        Dim attrs() As ExampleAttribute
        attrs = mi.GetAttributes(Of Example1Attribute)(False)
        Assert.IsTrue(attrs.Count = 1)
        Assert.AreEqual("classB", attrs(0).Message)

        attrs = mi.GetAttributes(Of Example1Attribute)(True)
        Assert.IsTrue(attrs.Count = 3)
        Assert.IsTrue(attrs.Any(Function(attr) attr.Message = "classAfirst"))
        Assert.IsTrue(attrs.Any(Function(attr) attr.Message = "classAsecond"))
        Assert.IsTrue(attrs.Any(Function(attr) attr.Message = "classB"))

        attrs = mi.GetAttributes(Of Example3Attribute)(True)
        Assert.IsTrue(attrs.Count = 0)
    End Sub

    <Test()>
    Public Sub HasAttributeTest()
        Dim mi = GetType(ClassB).GetMethod("Routine")

        Assert.IsFalse(mi.HasAttribute(Of Example2Attribute)(False))
        Assert.IsTrue(mi.HasAttribute(Of Example2Attribute)(True))
    End Sub

    <Test()>
    Public Sub GetAttribute_FailsWhenNonExistentTest()
        Dim mi = GetType(ClassB).GetMethod("Routine")

        Assert.Throws(Of InvalidOperationException)(
            Sub()
                mi.GetAttribute(Of Example2Attribute)(False)
            End Sub)

        Assert.Throws(Of InvalidOperationException)(
            Sub()
                mi.GetAttribute(Of Example3Attribute)(True)
            End Sub)
    End Sub

    <Test()>
    Public Sub GetAttribute_FailsWhenMultipleTest()
        Dim mi = GetType(ClassB).GetMethod("Routine")

        Assert.Throws(Of InvalidOperationException)(
            Sub()
                mi.GetAttribute(Of Example1Attribute)(True)
            End Sub)
    End Sub

    <Test()>
    Public Sub GetAttributeTest()
        Dim mi = GetType(ClassB).GetMethod("Routine")

        Dim attr As ExampleAttribute
        attr = mi.GetAttribute(Of Example1Attribute)(False)
        Assert.AreEqual("classB", attr.Message)
    End Sub

End Class

MustInherit Class ExampleAttribute
    Inherits Attribute

    Property Message As String

    Public Sub New(Optional msg As String = "")
        Message = msg
    End Sub

End Class


<AttributeUsage(AttributeTargets.All, AllowMultiple:=True, Inherited:=True)>
Class Example1Attribute
    Inherits ExampleAttribute

    Public Sub New(Optional msg As String = "")
        MyBase.New(msg)
    End Sub
End Class

<AttributeUsage(AttributeTargets.All, AllowMultiple:=True, Inherited:=True)>
Class Example2Attribute
    Inherits ExampleAttribute

    Public Sub New(Optional msg As String = "")
        MyBase.New(msg)
    End Sub
End Class

<AttributeUsage(AttributeTargets.All, AllowMultiple:=True, Inherited:=True)>
Class Example3Attribute
    Inherits ExampleAttribute

    Public Sub New(Optional msg As String = "")
        MyBase.New(msg)
    End Sub
End Class

Class ClassA

    <Example1("classAfirst"), Example1("classAsecond"), Example2("classA")>
    Public Overridable Sub Routine()

    End Sub

End Class


Class ClassB
    Inherits ClassA

    <Example1("classB")>
    Public Overrides Sub Routine()

    End Sub

End Class

File: Src/Icm.Core.Tests/Reflection/MethodInfoExtensionsTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using Icm.Reflection;
using NUnit.Framework;

[TestFixture()]
public class MethodInfoExtensionsTest
{

	[Test()]
	public void GetAttributesTest()
	{
		dynamic mi = typeof(ClassB).GetMethod("Routine");

		ExampleAttribute[] attrs = null;
		attrs = mi.GetAttributes<Example1Attribute>(false);
		Assert.That(attrs.Count == 1);
		Assert.AreEqual("classB", attrs(0).Message);

		attrs = mi.GetAttributes<Example1Attribute>(true);
		Assert.That(attrs.Count == 3);
		Assert.That(attrs.Any(attr => attr.Message == "classAfirst"));
		Assert.That(attrs.Any(attr => attr.Message == "classAsecond"));
		Assert.That(attrs.Any(attr => attr.Message == "classB"));

		attrs = mi.GetAttributes<Example3Attribute>(true);
		Assert.That(attrs.Count == 0);
	}

	[Test()]
	public void HasAttributeTest()
	{
		dynamic mi = typeof(ClassB).GetMethod("Routine");

		Assert.IsFalse(mi.HasAttribute<Example2Attribute>(false));
		Assert.That(mi.HasAttribute<Example2Attribute>(true));
	}

	[Test()]
	public void GetAttribute_FailsWhenNonExistentTest()
	{
		dynamic mi = typeof(ClassB).GetMethod("Routine");

		Assert.Throws<InvalidOperationException>(() => { mi.GetAttribute<Example2Attribute>(false); });

		Assert.Throws<InvalidOperationException>(() => { mi.GetAttribute<Example3Attribute>(true); });
	}

	[Test()]
	public void GetAttribute_FailsWhenMultipleTest()
	{
		dynamic mi = typeof(ClassB).GetMethod("Routine");

		Assert.Throws<InvalidOperationException>(() => { mi.GetAttribute<Example1Attribute>(true); });
	}

	[Test()]
	public void GetAttributeTest()
	{
		dynamic mi = typeof(ClassB).GetMethod("Routine");

		ExampleAttribute attr = null;
		attr = mi.GetAttribute<Example1Attribute>(false);
		Assert.AreEqual("classB", attr.Message);
	}

}

abstract class ExampleAttribute : Attribute
{

	public string Message { get; set; }

	public ExampleAttribute(string msg = "")
	{
		Message = msg;
	}

}


[AttributeUsage(AttributeTargets.All, AllowMultiple = true, Inherited = true)]
class Example1Attribute : ExampleAttribute
{

	public Example1Attribute(string msg = "") : base(msg)
	{
	}
}

[AttributeUsage(AttributeTargets.All, AllowMultiple = true, Inherited = true)]
class Example2Attribute : ExampleAttribute
{

	public Example2Attribute(string msg = "") : base(msg)
	{
	}
}

[AttributeUsage(AttributeTargets.All, AllowMultiple = true, Inherited = true)]
class Example3Attribute : ExampleAttribute
{

	public Example3Attribute(string msg = "") : base(msg)
	{
	}
}

class ClassA
{

	[Example1("classAfirst"), Example1("classAsecond"), Example2("classA")]

	public virtual void Routine()
	{
	}

}


class ClassB : ClassA
{

	[Example1("classB")]

	public override void Routine()
	{
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
